README.rst
Mopidy-Bandcamp
****************

.. image:: https://img.shields.io/pypi/v/Mopidy-Bandcamp?logo=PyPI
    :target: https://pypi.org/project/Mopidy-Bandcamp
    :alt: Latest PyPI version

.. image:: https://img.shields.io/github/v/release/impliedchaos/mopidy-bandcamp?logo=GitHub
    :target: https://github.com/impliedchaos/mopidy-bandcamp/releases
    :alt: Latest GitHub release

.. image:: https://img.shields.io/github/commits-since/impliedchaos/mopidy-bandcamp/latest?logo=GitHub
    :target: https://github.com/impliedchaos/mopidy-bandcamp/commits/master
    :alt: Latest GitHub Commits

.. image:: https://img.shields.io/github/workflow/status/impliedchaos/mopidy-bandcamp/CI?logo=GitHub&label=CI%20build
    :target: https://github.com/impliedchaos/mopidy-bandcamp/actions
    :alt: CI build status

.. image:: https://img.shields.io/codecov/c/github/impliedchaos/mopidy-bandcamp?logo=Codecov&label=Test%20coverage
    :target: https://app.codecov.io/gh/impliedchaos/mopidy-bandcamp/
    :alt: Code Coverage

.. image:: https://img.shields.io/badge/PRs-welcome-brightgreen
    :target: https://makeapullrequest.com/
    :alt: Pull Requests Welcome


`Mopidy <http://www.mopidy.com/>`_ extension for playing music from
`Bandcamp <http://bandcamp.com/>`_.

This backend allows searching bandcamp and playing the free 128kbps MP3 streams.

Initial support has been added (as of v1.1.0) for browsing and playing your bandcamp collection.
Authentication is a hassle, and described below.  Expect things to be wonky, and
please create an issue when you encounter things that don't work.  Also this is slow
because it requires scraping the bandcamp website instead of using an API.
(This plugin will attempt to play the higher quality mp3-v0 streams of songs in your collection.)


Installation
============

Install by running::

    sudo python3 -m pip install Mopidy-Bandcamp


Authentication
==============

If you wish to browse/play music in your collection, then you'll need to authenticate to bandcamp.com.

Authentication is done by grabbing your :code:`identity` token from the cookies of the
bandcamp website. Point your browser at https://bandcamp.com and log in if you aren't already.

For chrome you can simply go to 
`chrome://settings/cookies/detail?site=bandcamp.com <chrome://settings/cookies/detail?site=bandcamp.com>`_
to see the cookies for bandcamp.com.  Click on "identity" and copy the data listed under "Content".

For Firefox and most other browsers, open up the browser's developer tools (usually by pressing Ctrl-Shift-I or F12).
When the tools pop up, navigate to the "Storage" tab, and then select "Cookies" to view
the site's cookies. Copy the value from the "identity" cookie.

It should look something like this:

.. code::

    7%09xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx%3D%09%7B%22ex%22%3Ax%2C%22id%22%3Axxxxxxxxxx%7D

Now add it to your Mopidy config file like so:

.. code::

    [bandcamp]
    identity = 7%09xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx%3D%09%7B%22ex%22%3Ax%2C%22id%22%3Axxxxxxxxxx%7D


Usage
=====

Albums and tracks can be searched for or browsed.  Additionaly, you can force Mopidy-Bandcamp to scrape
a bandcamp URL for you by just prepending the url with "bandcamp:" and adding it to Mopidy.  For example,
using mpc on the command line:

.. code:: shell

    $ mpc add "bandcamp:https://louiezong.bandcamp.com/album/jazz"

Would add the bandcamp album to the queue.

Note: Adding an artist by clicking on the artist in a search result or by manually scraping an artist's
bandcamp page can take a long time depending on the artist.  This is because Mopidy-Bandcamp tries to load
the entirety of the artist's discography.


Web Client
----------

As of v1.1.2 a simple webclient has been added to allow for more easily scraping a page. Not by going to
http://hostname:6680/bandcamp/ and entering in a url (which you can do), but by using the following
as a URL for a bookmark in your web browser:

.. code::

    javascript:s='http://hostname:6680/bandcamp/';f=document.createElement('form');f.action=s;i=document.createElement('input');i.type='hidden';i.name='url';i.value=window.location.href;f.appendChild(i);document.body.appendChild(f);f.submit();

Note: Replace *hostname* and *6680* with your mopidy server's hostname and configured HTTP port.

Now when you're browsing bandcamp you can simply click that bookmark to add the current page to Mopidy.
(This works in Chrome and Firefox.  I haven't bothered checking anything else.)

Configuration
=============

example:

.. code::

    [bandcamp]
    discover-tags = French House, Brit Pop, Tokyo, New Wave, Industrial


- :code:`identity` - Identity token from your bandcamp cookies to authenticate with Bandcamp.
- :code:`collection_items` - Number of items (per page) to fetch from your collection (if authenticated).  Default: 50
- :code:`discover_tags` - List of tags to discover. **You'll really want to change this to any tags that interest you.**
- :code:`discover_genres` - List of bandcamp's genres to discover.  You'll only want to edit this to remove unwanted genres.
- :code:`discover_pages` - Number of pages to load in the browse discover sections.  Default: 1
- :code:`image_sizes` - a list of ids for image sizes to return to mopidy for album art.  Default: 10, 5, 2 (1200x1200, 700x700, 350x350)


Bandcamp image size ids:

+----+-------+--------+--------+
| ID | Width | Height | Aspect |
+====+=======+========+========+
| 1  | Original (usually big)  |
+----+-------+--------+--------+
| 10 | 1200  | 1200   | Square |
+----+-------+--------+--------+
| 20 | 1024  | 1024   | Square |
+----+-------+--------+--------+
| 5  | 700   | 700    | Square |
+----+-------+--------+--------+
| 13 | 380   | 380    | Square |
+----+-------+--------+--------+
| 14 | 368   | 368    | Square |
+----+-------+--------+--------+
| 2  | 350   | 350    | Square |
+----+-------+--------+--------+
| 4  | 300   | 300    | Square |
+----+-------+--------+--------+
| 9  | 210   | 210    | Square |
+----+-------+--------+--------+
| 44 | 200   | 200    | Square |
+----+-------+--------+--------+
| 11 | 172   | 172    | Square |
+----+-------+--------+--------+
| 7  | 150   | 150    | Square |
+----+-------+--------+--------+
| 50 | 140   | 140    | Square |
+----+-------+--------+--------+
| 12 | 138   | 138    | Square |
+----+-------+--------+--------+
| 15 | 135   | 135    | Square |
+----+-------+--------+--------+
| 8  | 124   | 124    | Square |
+----+-------+--------+--------+
| 21 | 120   | 120    | Square |
+----+-------+--------+--------+
| 3  | 100   | 100    | Square |
+----+-------+--------+--------+
| 42 | 50    | 50     | Square |
+----+-------+--------+--------+
| 22 | 25    | 25     | Square |
+----+-------+--------+--------+
| 26 | 800   | 600    | 4:3    |
+----+-------+--------+--------+
| 36 | 400   | 300    | 4:3    |
+----+-------+--------+--------+
| 32 | 380   | 285    | 4:3    |
+----+-------+--------+--------+
| 33 | 368   | 276    | 4:3    |
+----+-------+--------+--------+
| 37 | 168   | 126    | 4:3    |
+----+-------+--------+--------+
| 38 | 144   | 108    | 4:3    |
+----+-------+--------+--------+
| 29 | 100   | 75     | 4:3    |
+----+-------+--------+--------+
| 28 | 768   | 432    | 16:9   |
+----+-------+--------+--------+
| 27 | 715   | 402    | 16:9   |
+----+-------+--------+--------+


Project resources
=================

- `Source code <https://github.com/impliedchaos/mopidy-bandcamp>`_
- `Issue tracker <https://github.com/impliedchaos/mopidy-bandcamp/issues>`_


CHANGELOG.md
Change Log
==========

v1.1.5
------

- fixed another bug in scrape function for unauthenticated accounts
- added Wishlist to browsing directory

v1.1.4
------

- Quick documentation update.
- Fix for a change in Bandcamp's JSON formatting.

v1.1.3
------

- added dummy images for browse directories.
- better handling of weird calls to library.lookup by front-ends like Iris
- fixed bug in scrape function for unauthenticated accounts

v1.1.2
------

- added webclient for adding bandcamp urls from the browser
- small bug fixes, added edge test cases

v1.1.1
------

- added support for scraping bandcamp websites

v1.1.0
------

- added support for auth via cookie
- added support for browsing collection
- removed "art_url_as_comment" config option

v1.0.1
------

- bug fixes after implenting more edge test cases

v1.0.0
------

YAY!

- updated error handling
- added testing and code coverage

v0.2.1
------

- added image size preferences

v0.2.0
------

- added browsing by tag or genre
- code cleanup/refactoring

v0.1.1
------

- bug fixes
- code refactoring

v0.1.0
------

- initial release


tests/test_web.py
from unittest import mock

import tornado.testing
import tornado.web
import tornado.websocket

import mopidy.config as config
from mopidy_bandcamp import Extension


class BaseTest(tornado.testing.AsyncHTTPTestCase):
    def get_app(self):
        extension = Extension()
        self.config = config.Proxy({})
        return tornado.web.Application(extension.factory(self.config, mock.Mock()))


class WebHandlerTest(BaseTest):
    def test_index(self):
        response = self.fetch("/", method="GET")
        assert response.code == 200
        response = self.fetch("/?url=https%3A%2F%2Fgoogle.com%2F", method="GET")
        assert response.code == 200
        body = tornado.escape.to_unicode(response.body)
        assert "<title>Error</title>" in body
        response = self.fetch(
            "/?url=https%3A%2F%2Flouiezong.bandcamp.com%2Ftrack%2Fbrain-age",
            method="GET",
        )
        assert response.code == 200
        body = tornado.escape.to_unicode(response.body)
        assert "<title>URL Added</title>" in body


tests/test_extension.py
import os
import unittest
from unittest import mock
from time import sleep

from mopidy.models import Track, SearchResult, Image
from mopidy_bandcamp import Extension
from mopidy_bandcamp import backend as backend_lib


class ExtensionTest(unittest.TestCase):
    @staticmethod
    def get_config():
        config = {}
        config["enabled"] = True
        config["discover_pages"] = 1
        config["discover_genres"] = [
            "All",
            "Electronic",
            "Rock",
            "Metal",
            "Alternative",
            "Hip-Hop/Rap",
            "Experimental",
            "Punk",
            "Folk",
            "Pop",
            "Ambient",
            "Soundtrack",
            "World",
            "Jazz",
            "Acoustic",
            "Funk",
            "R&B/Soul",
            "Devotional",
            "Classical",
            "Reggae",
            "Podcasts",
            "Country",
            "Spoken Word",
            "Comedy",
            "Blues",
            "Kids",
            "Audiobooks",
            "Latin",
        ]
        config["discover_tags"] = [
            "Outrun",
            "Future Funk",
            "Alternative Hip-Hop",
            "Tokyo, Japan",
        ]
        config["image_sizes"] = ["10", "5", "2"]
        config["identity"] = ""
        config["collection_items"] = 50
        return {"bandcamp": config, "proxy": {}}

    def test_get_default_config(self):
        ext = Extension()

        config = ext.get_default_config()

        assert "[bandcamp]" in config
        assert "enabled=true" in config

    def test_get_config_schema(self):
        ext = Extension()
        schema = ext.get_config_schema()

        assert "enabled" in schema
        assert "discover_pages" in schema
        assert "discover_genres" in schema
        assert "discover_tags" in schema
        assert "image_sizes" in schema
        assert "identity" in schema
        assert "collection_items" in schema

    def test_get_backend_classes(self):
        registry = mock.Mock()
        ext = Extension()
        ext.setup(registry)
        assert (
            mock.call("backend", backend_lib.BandcampBackend) in registry.add.mock_calls
        )

    def test_init_backend(self):
        backend = backend_lib.BandcampBackend(ExtensionTest.get_config(), None)
        assert backend is not None
        backend.on_start()
        backend.on_stop()

    def test_browse(self):
        backend = backend_lib.BandcampBackend(ExtensionTest.get_config(), None)
        root = backend.library.browse(None)
        assert root == []
        root = backend.library.browse("chewbacca")
        assert root is None
        root = backend.library.browse("bandcamp:browse")
        assert len(root) == 2
        g1 = backend.library.browse("bandcamp:genres")
        assert len(g1) == len(backend.library.genres)
        assert g1[1].name == "Electronic"
        g2 = backend.library.browse(g1[1].uri + ":2")
        # Length SHOULD be 49, but Bandcamp is friggin crazy, and doesn't
        # always send the same number of items.
        assert len(g2) > 40
        t1 = backend.library.browse("bandcamp:tags")
        assert len(t1) == len(backend.library.tags)
        assert t1[1].name == "Future Funk"
        t2 = backend.library.browse(t1[1].uri)
        # Length SHOULD be 49, but Bandcamp is friggin crazy, and doesn't
        # always send the same number of items.
        assert len(t2) > 40
        a = backend.library.browse(t2[0].uri)
        # Album browse
        assert len(a) > 1

    def test_search_lookup_translate(self):
        backend = backend_lib.BandcampBackend(ExtensionTest.get_config(), None)
        track = backend.library.lookup("bandcamp:track:nope-0-nuhuh")
        assert track == []
        artist = backend.library.lookup("bandcamp:artist:nope")
        assert artist == []
        album = backend.library.lookup("bandcamp:album:nope-nuhuh")
        assert album == []
        sleep(1)
        res = backend.library.search(
            {"any": "Waveshaper", "album": ["Station", "Nova"]}
        )
        assert isinstance(res, SearchResult)
        res = backend.library.search(["Station", "Nova"])
        assert isinstance(res, SearchResult)
        res = backend.library.search("Waveshaper")
        assert isinstance(res, SearchResult)
        artist = backend.library.lookup("bandcamp:nothing")
        assert artist == []
        sleep(1)
        artist = backend.library.lookup("bandcamp:artist:4274249518")
        assert isinstance(artist[0], Track)
        album = backend.library.lookup("bandcamp:album:4274249518-4240848302")
        assert isinstance(album[0], Track)
        sleep(0.5)
        track = backend.library.lookup("bandcamp:track:1715582766-4294604529-839491036")
        assert isinstance(track[0], Track)
        track = backend.library.lookup("bandcamp:https://jasonwebley.bandcamp.com/track/in-this-light")
        assert isinstance(track[0], Track)
        assert track[0].uri == 'bandcamp:track:2220410354-3763840907-1951117791'
        img = backend.library.get_images(
            ["bandcamp:track:1715582766-4294604529-839491036"]
        )
        assert isinstance(
            img["bandcamp:track:1715582766-4294604529-839491036"][0], Image
        )
        img = backend.library.get_images(["bandcamp:browse", "bandcamp:tag:outrun"])
        assert isinstance(img["bandcamp:browse"][0], Image)
        assert isinstance(img["bandcamp:tag:outrun"][0], Image)
        uri = backend.playback.translate_uri(album[0].uri)
        assert uri.startswith("http")
        uri = backend.playback.translate_uri("chewbacca")
        assert uri is None
        uri = backend.playback.translate_uri(
            "bandcamp:mytrack:1113572798-2944001033-1139398039"
        )
        assert uri is None
        uri = backend.playback.translate_uri("bandcamp:track:0-0-0")
        assert uri is None

    def test_other(self):
        cfg = ExtensionTest.get_config()
        cfg["bandcamp"]["art_url_as_comment"] = True
        cfg["bandcamp"]["discover_pages"] = 0
        cfg["bandcamp"]["image_sizes"] = ["1"]
        backend = backend_lib.BandcampBackend(cfg, None)
        res = backend.library.search("Waveshaper")
        assert isinstance(res, SearchResult)
        track = backend.library.lookup("bandcamp:track:4274249518-4240848302-55800693")
        assert isinstance(track[0], Track)
        img = backend.library.get_images(
            ["bandcamp:track:4274249518-4240848302-55800693"]
        )
        assert isinstance(
            img["bandcamp:track:4274249518-4240848302-55800693"][0], Image
        )
        root = backend.library.browse("bandcamp:browse")
        assert root == []
        root = backend.library.browse("bandcamp:collection:0:0:a::")
        assert root == []
        root = backend.library.browse("bandcamp:wishlist:0:0:a::")
        assert root == []

    def test_auth(self):
        cfg = ExtensionTest.get_config()
        cfg["bandcamp"]["collection_items"] = 2
        cfg["bandcamp"]["identity"] = os.environ.get("TEST_BANDCAMP_ID")
        if cfg["bandcamp"]["identity"] is not None:
            backend = backend_lib.BandcampBackend(cfg, None)
            root = backend.library.browse("bandcamp:browse")
            assert len(root) == 4
            col = backend.library.browse("bandcamp:collection")
            assert len(col) >= 1
            wish = backend.library.browse("bandcamp:wishlist")
            assert len(wish) >= 1
            alb = backend.library.browse(col[0].uri)
            assert len(alb) >= 1
            # Track we own
            track = backend.library.lookup("bandcamp:mytrack:2937093478-0-4173304455")
            assert isinstance(track[0], Track)
            uri = backend.playback.translate_uri(
                "bandcamp:mytrack:2937093478-0-4173304455"
            )
            assert uri.startswith("http")
            # Free track we don't own
            track = backend.library.lookup("bandcamp:mytrack:2937093478-0-1532026871")
            assert isinstance(track[0], Track)
            uri = backend.playback.translate_uri(
                "bandcamp:mytrack:2937093478-0-1532026871"
            )
            assert uri.startswith("http")
            uri = backend.playback.translate_uri("bandcamp:mytrack:nope-0-notgonnawork")
            assert uri is None

    def test_scrape(self):
        cfg = ExtensionTest.get_config()
        backend = backend_lib.BandcampBackend(cfg, None)
        nothing = backend.library.lookup("bandcamp:https://bandcamp.com/")
        assert nothing == []
        artist = backend.library.lookup("bandcamp:https://kitschclub.bandcamp.com/")
        assert len(artist) > 1
        album = backend.library.lookup(
            "bandcamp:https://louiezong.bandcamp.com/album/jazz"
        )
        assert len(album) > 1
        track = backend.library.lookup(
            "bandcamp:https://louiezong.bandcamp.com/track/minty-fresh"
        )
        assert isinstance(track[0], Track)

    def test_scrape_auth(self):
        cfg = ExtensionTest.get_config()
        cfg["bandcamp"]["identity"] = os.environ.get("TEST_BANDCAMP_ID")
        backend = backend_lib.BandcampBackend(cfg, None)
        nothing = backend.library.lookup("bandcamp:https://bandcamp.com/")
        assert nothing == []
        artist = backend.library.lookup("bandcamp:https://kitschclub.bandcamp.com/")
        assert len(artist) > 1
        album = backend.library.lookup(
            "bandcamp:https://louiezong.bandcamp.com/album/jazz"
        )
        assert len(album) > 1
        track = backend.library.lookup(
            "bandcamp:https://louiezong.bandcamp.com/track/minty-fresh"
        )
        assert isinstance(track[0], Track)


mopidy_bandcamp/web.py
import re
import tornado.web


class WebHandler(tornado.web.RequestHandler):
    def initialize(self, config, core):
        self.core = core
        self.config = config

    def get(self):
        url = self.get_argument("url", None)
        if url is not None:
            if re.match(r"^https?://[^/]+.bandcamp\.com/", url):
                self.core.tracklist.add(uris=[f"bandcamp:{url}"])
                self.write(
                    "<!DOCTYPE html><html><head><title>URL Added</title><script>alert('URL has been added.');window.history.back()</script></head></html>"
                )
            else:
                self.write(
                    "<!DOCTYPE html><html><head><title>Error</title></head><body><h1>Error</h1><p>Invalid URL: &quot;%s&quot;</p></body></html>"
                    % url
                )
        else:
            self.write(
                """<!DOCTYPE html>
<html><head><title>Add a bandcamp URL</title></head>
<body><div><h1>Add a bandcamp URL to Mopidy tracklist</h1>
<form><p>URL: <input name="url" type="text" /></p><p><input type="submit" /></p></form><p></div>
<div><h1>Why is this here?</h1>
<p>The purpose of this is to be able to easily add bandcamp albums to Mopidy from your web browser.</p>
<textarea id="bookmark" rows="4" cols="80"></textarea>
<script>document.getElementById('bookmark').value="javascript:s='" + window.location +
"';f=document.createElement('form');f.action=s;i=document.createElement('input');i.type='hidden';" +
"i.name='url';i.value=window.location.href;f.appendChild(i);document.body.appendChild(f);f.submit();"
</script>
<p>Copy the above snippet, and create a bookmark in your web browser with the snippet as the URL.<br/>
When you're browsing bandcamp, you can simply click that bookmark to add the current page to Mopidy.</p>
<a href="https://github.com/impliedchaos/mopidy-bandcamp#web-client">more info</a></div></body></html>
"""
            )


mopidy_bandcamp/backend.py
import pykka
import re

from mopidy import backend
from mopidy_bandcamp import logger
from mopidy_bandcamp.bandcamp import BandcampClient
from mopidy_bandcamp.library import BandcampLibraryProvider


class BandcampBackend(pykka.ThreadingActor, backend.Backend):
    def __init__(self, config, audio):
        super(BandcampBackend, self).__init__()
        self.config = config
        self.audio = audio
        self.uri_schemes = ["bandcamp"]
        self.image_sizes = []
        for s in config["bandcamp"]["image_sizes"]:
            self.image_sizes.append(str(s))
        self.bandcamp = BandcampClient(config)
        self.library = BandcampLibraryProvider(backend=self)
        self.playback = BandcampPlaybackProvider(audio=audio, backend=self)


class BandcampPlaybackProvider(backend.PlaybackProvider):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def translate_uri(self, uri):
        logger.debug('"Bandcamp translate_uri "%s"', uri)
        if re.match(r"^bandcamp:(my)?track:", uri) is None:
            return None
        try:
            _, func, p = uri.split(":")
            artist, album, track = p.split("-")
            if func == "mytrack":
                if uri in self.backend.library.scrape_urls:
                    logger.debug("Bandcamp found url in track scrape url cache")
                    url = self.backend.library.scrape_urls[uri]
                elif p in self.backend.library.tracks and self.backend.library.tracks[
                    p
                ].comment.startswith("URL: "):
                    logger.debug("Bandcamp found url in track comment")
                    url = self.backend.library.tracks[p].comment.split(": ")[1]
                else:
                    logger.debug("Bandcamp fetching url from web")
                    resp = self.backend.bandcamp.get_track(artist, track)
                    url = resp["bandcamp_url"]
                resp = self.backend.bandcamp.scrape(url)
                for tr in resp["tracks"]:
                    if str(tr["id"]) == track:
                        if "mp3-v0" in tr["file"]:
                            logger.debug(
                                "Bandcamp found mp3-v0 url '%s'",
                                tr["file"]["mp3-v0"],
                            )
                            return tr["file"]["mp3-v0"]
                        else:
                            if "mp3-128" in tr["file"]:
                                logger.debug(
                                    "Bandcamp found mp3-128 url '%s'",
                                    tr["file"]["mp3-128"],
                                )
                                return tr["file"]["mp3-128"]
                logger.error("Track not found. %s", uri)  # pragma: no cover
            else:
                resp = self.backend.bandcamp.get_track(artist, track)
                if "tracks" in resp:
                    tr = resp["tracks"][0]
                    if "streaming_url" in tr and "mp3-128" in tr["streaming_url"]:
                        logger.debug(
                            "Bandcamp found url '%s'",
                            tr["streaming_url"]["mp3-128"],
                        )
                        return tr["streaming_url"]["mp3-128"]
                logger.error("Track not found. %s", uri)  # pragma: no cover
        except Exception as e:
            logger.error('translate_uri error "%s"', str(e))
        return None


mopidy_bandcamp/__init__.py
import logging
import os

from mopidy import config, ext

__version__ = "1.1.5"

logger = logging.getLogger(__name__)


class Extension(ext.Extension):

    dist_name = "Mopidy-Bandcamp"
    ext_name = "bandcamp"
    version = __version__

    def get_default_config(self):
        conf_file = os.path.join(os.path.dirname(__file__), "ext.conf")
        return config.read(conf_file)

    def get_config_schema(self):
        schema = super(Extension, self).get_config_schema()
        schema["discover_pages"] = config.Integer(optional=True)
        schema["collection_items"] = config.Integer(optional=True)
        schema["discover_genres"] = config.List(optional=True)
        schema["discover_tags"] = config.List(optional=True)
        schema["image_sizes"] = config.List(optional=True)
        schema["identity"] = config.String(optional=True)
        return schema

    def setup(self, registry):
        from .backend import BandcampBackend

        registry.add("backend", BandcampBackend)
        registry.add("http:app", {"name": self.ext_name, "factory": self.factory})

    def factory(self, config, core):
        from .web import WebHandler

        return [(r"/", WebHandler, {"config": config, "core": core})]


mopidy_bandcamp/bandcamp.py
import json
import mopidy_bandcamp
import re
import requests

from html import unescape
from mopidy import httpclient
from time import time


class BandcampClient:
    BASE_URL = "https://bandcamp.com/api"
    PAGE_ITEMS = 48
    IMAGE_SIZE = {
        "10": [1200, 1200],
        "20": [1024, 1024],
        "5": [700, 700],
        "13": [380, 380],
        "14": [368, 368],
        "2": [350, 350],
        "4": [300, 300],
        "9": [210, 210],
        "44": [200, 200],
        "11": [172, 172],
        "7": [150, 150],
        "50": [140, 140],
        "12": [138, 138],
        "15": [135, 135],
        "8": [124, 124],
        "21": [120, 120],
        "3": [100, 100],
        "42": [50, 50],
        "22": [25, 25],
        "26": [800, 600],
        "36": [400, 300],
        "32": [380, 285],
        "33": [368, 276],
        "37": [168, 126],
        "38": [144, 108],
        "29": [100, 75],
        "28": [768, 432],
        "27": [715, 402],
    }

    def __init__(self, config):
        self.proxy = httpclient.format_proxy(config["proxy"])
        self.ua_str = httpclient.format_user_agent(
            f"{mopidy_bandcamp.Extension.dist_name}/{mopidy_bandcamp.__version__}"
        )
        self.identity = config["bandcamp"]["identity"]
        self.collection_items = config["bandcamp"]["collection_items"]
        self.fan_id = None

    def _get(self, *args, **kwargs):
        headers = {"User-Agent": self.ua_str}
        resp = requests.get(*args, **kwargs, headers=headers, proxies=self.proxy)
        resp.raise_for_status()
        js = resp.json()
        if "error" in js:
            raise RuntimeError(js["error_message"])
        return js

    def _post(self, *args, **kwargs):
        headers = {"User-Agent": self.ua_str}
        resp = requests.post(*args, **kwargs, headers=headers, proxies=self.proxy)
        resp.raise_for_status()
        js = resp.json()
        if "error" in js:
            raise RuntimeError(js["error_message"])
        return js

    def scrape(self, uri):
        headers = {"User-Agent": self.ua_str}
        if self.identity:
            headers["Cookie"] = f"identity={self.identity}"
        resp = requests.get(uri, headers=headers, proxies=self.proxy)
        resp.raise_for_status()
        # Build the tralbum data by joining multiple json chunks.
        data = re.search(r'\s+data-tralbum="(.*?)"', resp.text)
        if data is None:
            raise RuntimeError("Couldn't scrape data-tralbum from " + uri)
        tralbum = json.loads(unescape(data.group(1)))
        if "trackinfo" not in tralbum:
            # Artist page.
            data = re.search(r'\s+data-band="(.*?)"', resp.text)
            if data is None:
                raise RuntimeError(
                    "Couldn't scrape data-band from " + uri
                )  # pragma: no cover
            tralbum.update(json.loads(unescape(data.group(1))))
        else:
            # Album/track page.
            data = re.search(r'\s+data-band-follow-info="(.*?)"', resp.text)
            if data is None:
                raise RuntimeError(
                    "Couldn't scrape data-band-follow-info from " + uri
                )  # pragma: no cover
            tralbum.update(json.loads(unescape(data.group(1))))
            data = re.search(r'\s+data-embed="(.*?)"', resp.text)
            if data is None:
                raise RuntimeError(
                    "Couldn't scrape data-embed from " + uri
                )  # pragma: no cover
            tralbum.update(json.loads(unescape(data.group(1))))
            if "band_id" not in tralbum:
                if "current" in tralbum and "band_id" in tralbum["current"]:
                    tralbum["band_id"] = tralbum["current"]["band_id"]
                else:
                    tralbum["band_id"] = None
            data = re.search(r'application/ld\+json">\s*(.*?)\s*</script', resp.text)
            if data is not None:
                ld = json.loads(data.group(1))
                tralbum["keywords"] = ld["keywords"]
                if tralbum["band_id"] is None:
                    if "byArtist" in ld and "additionalProperty" in ld["byArtist"]:
                        for ap in ld["byArtist"]["additionalProperty"]:
                            if ap["name"] == "band_id":
                                tralbum["band_id"] = ap["value"]
                                break
            tralbum["tracks"] = tralbum["trackinfo"]
            tralbum["tralbum_artist"] = tralbum["artist"]
            tralbum["num_downloadable_tracks"] = None
        return tralbum

    def get_collection(self, token=None, ctype="collection"):
        if not self.identity:
            return []
        headers = {"User-Agent": self.ua_str, "Cookie": f"identity={self.identity}"}
        if self.fan_id is None:
            resp = requests.get(
                self.BASE_URL + "/fan/2/collection_summary",
                headers=headers,
                proxies=self.proxy,
            )
            resp.raise_for_status()
            self.fan_id = resp.json()["fan_id"]
        if token is None:
            token = str(time()) + ":0:a::"
        endpoint = "collection_items"
        if ctype == "wishlist":
            endpoint = "wishlist_items"
        elif ctype == "following":
            endpoint = "following_bands"
        resp = requests.post(
            self.BASE_URL + "/fancollection/1/" + endpoint,
            headers=headers,
            proxies=self.proxy,
            json={
                "fan_id": self.fan_id,
                "older_than_token": token,
                "count": self.collection_items,
            },
        )
        resp.raise_for_status()
        js = resp.json()
        if "error" in js:
            raise RuntimeError(js["error_message"])  # pragma: no cover
        return js

    def get_album(self, artistid, itemid, track=False):
        js = self._get(
            self.BASE_URL
            + "/mobile/24/tralbum_details?band_id="
            + str(artistid)
            + "&tralbum_type="
            + ("t" if track else "a")
            + "&tralbum_id="
            + str(itemid)
        )
        return js

    def get_track(self, artistid, trackid):
        return self.get_album(artistid, trackid, track=True)

    def get_artist(self, artistid):
        js = self._post(
            self.BASE_URL + "/mobile/24/band_details",
            json={"band_id": artistid},
        )
        return js

    def search(self, query):
        js = self._get(f"{self.BASE_URL}/fuzzysearch/1/app_autocomplete?q={query}")
        return js

    def discover(self, sort="top", genre="all", tag=None, page=0):
        query = f"{self.BASE_URL}/discover/2/get?s={sort}&l=0&emulate_loc=true&f=all&g={genre}"
        if tag is not None:
            query += f"&t={tag}"
        if page > 0:
            query += f"&p={page}"
        js = self._get(query)
        return js
